Update stats in process

russelldb opened the pull request:
When a node is giving good thoughput then riak_kv_stat cannot keep
up. Large message queues ensue, memory balloons, the node dies.
This commit only forces stat creation through the gen_server, any
other stat update is direct in process.

This does cause a degradation in throughput & latency as each get / put
process is waiting to acquire the locks on the folsom tables.

NOTE: this works a whole lot better with these `folsom` optimisations https://github.com/basho/folsom/tree/adt-faster-spiral

With the `folsom` improvements, put throughput is as good as 1.2.1 (though worse than master without stats.) Even with the `folsom` improvements, there is extra latency, most noticeable in the 95th%ile.

russelldb committed:
Update stats in process

When a node is giving good thoughput then riak_kv_stat cannot keep
up. Large message queues ensue, memory balloons, the node dies.
This commit only forces stat creation through the gen_server, any
other stat update is direct in process.

This does cause a degradation in throughput as each get / put
process is waiting to acquire the locks on the folsom tables.

russelldb committed:
Spawn a monitoring process per get/ put fsm

Un-gen_serverize get_put_monitor
Move the get/put stats to riak stat
Remove unused all_stats/0 fun from get/put monitor
Update eqc test to work with a spawn per fsm

joedevivo commented:
./riak_test -t verify_riak_stats -> pass

engelsanchez commented:
The change is simple looks good as far as I can tell. :+1: Was that your new plus one @joedevivo ?

joedevivo commented on src/riak_kv_stat.erl:
```diff
@@ -85,8 +99,9 @@ init([]) ->
+    {reply, Rep, State}.
 
 handle_cast({update, Arg}, State) ->
     do_update(Arg),
```
You should remove this handle_cast clause, since it's no longer used.

Commit:
riak_kv_stat: removed unused handle_call({update, Arg}) clause

joedevivo commented:
@engelsanchez no, that was just saying the code passed the tests, still was awaiting manual review. I told @russelldb  I'd remove the extra handle_cast and run riak_test again.

joedevivo commented:
Ok, riak_test passed again after the clause was removed. +1

russelldb committed:
Fix eqc tests

The previous test used processes to simulate a gen_server,
they expected casts / calls to riak_kv_get_put_monitor.
Use meck instead to simulate the interface of the get/put
monitor.

russelldb committed:
Fix typo in comments

russelldb committed:
Merge branch 'rdb_kv_stat_direct' of github.com:basho/riak_kv into rdb_kv_stat_direct

Conflicts:
	src/riak_kv_stat.erl
